src/wasm/wasm-ir-builder.cpp

```diff
@@ -1035,11 +1031,24 @@ Result<> IRBuilder::visitEnd() {
   } else if (auto* loop = scope.getLoop()) {
     loop->body = *expr;
     loop->name = scope.label;
+    if (scope.inputType != Type::none && scope.labelUsed) {
+      // Branches to this loop carry values, but Binaryen IR does not support
+      // that. Fix this by trampolining the branches through new code that sets
+      // the branch value to the appropriate scratch local.
+      fixLoopWithInput(loop, scope.inputType, scope.inputLocal);
+    }
     loop->finalize(loop->type);
     push(loop);
   } else if (auto* iff = scope.getIf()) {
     iff->ifTrue = *expr;
-    iff->ifFalse = nullptr;
+    if (scope.inputType != Type::none) {
+      // Normally an if without an else must have type none, but if there is an
+      // input parameter, the empty else arm must propagate its value.
+      // Synthesize an else arm that loads the value from the scratch local.
+      iff->ifFalse = builder.makeLocalGet(scope.inputLocal, scope.inputType);
```

Comment:
I don't follow this. There is an input type, but this makes the second arm have an output type. Can this arm simply not read the scratch local?

Reply:
Any if without an else must have equivalent input and output types.

In the specified WebAssembly semantics, there's actually no such thing as an if without an else arm. There are only ifs with empty else arms instead. It might be worth looking into removing the ability for `ifFalse` arms to be null in Binaryen IR; that might be a significant simplification.

Reply:
Oh, I see... right, if the input isn't read in an arm, it has to flow out.

Reply:
About disallowing null in our IR, interesting idea. It would have some memory usage downsides though. I don't have a good guess if it would help enough to justify that.